crates/boojum/src/gadgets/u2048/mod.rs

```diff
@@ -0,0 +1,485 @@
+            .zip(other.inner.iter())
+            .zip(result.inner.iter_mut())
+        {
+            let (c, borrow) = (*a).overflowing_sub_with_borrow_in(cs, *b, borrow_out);
+            *dst = c;
+            borrow_out = borrow;
+        }
+
+        (result, borrow_out)
+    }
+
+    #[must_use]
+    pub fn widening_mul<CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        other: &Self,
+        self_limbs: usize,
+        other_limbs: usize,
+    ) -> UInt4096<F> {
+        assert!(
+            self_limbs + other_limbs <= 128,
+            "total number of limbs must be <= 128"
+        );
+
+        let zero = UInt32::allocated_constant(cs, 0);
+        let mut remainders = vec![UInt32::<F>::zero(cs); self_limbs + other_limbs];
+
+        for i in 0..self_limbs {
+            let mut carry = UInt32::allocated_constant(cs, 0);
+            for j in 0..other_limbs {
+                let res = UInt32::fma_with_carry(
+                    cs,
+                    self.inner[i],
+                    other.inner[j],
+                    if i == 0 { zero } else { remainders[i + j] },
+                    carry,
+                );
+                (remainders[i + j], carry) = (res[0].0, res[1].0);
+            }
+            remainders[i + other_limbs] = carry;
+        }
+
+        let mut inner = [UInt32::<F>::zero(cs); 128];
+        inner[..self_limbs + other_limbs].copy_from_slice(&remainders);
+        UInt4096 { inner }
+    }
+
+    /// Multiplies a number by 2^{32}. Panics if the number overflows.
+    #[must_use]
+    pub fn must_mul_by_two_pow_32<CS: ConstraintSystem<F>>(&self, cs: &mut CS) -> Self {
+        let boolean_true = Boolean::allocated_constant(cs, true);
+        let last_limb_zero = self.inner[63].is_zero(cs);
+        Boolean::enforce_equal(cs, &last_limb_zero, &boolean_true);
+
+        let mut new_inner = self.inner;
+        new_inner.copy_within(0..65, 1);
+        new_inner[0] = UInt32::zero(cs);
+
+        Self { inner: new_inner }
+    }
+
+    #[must_use]
+    pub fn div2<CS: ConstraintSystem<F>>(&self, cs: &mut CS) -> Self {
+        let byte_split_id = cs
+            .get_table_id_for_marker::<ByteSplitTable<1>>()
+            .expect("table should exist");
+        let mut bytes = self.to_le_bytes(cs);
+        let mut bit: Option<Variable> = None;
+        bytes.iter_mut().rev().for_each(|b| {
+            let res = cs.perform_lookup::<1, 2>(byte_split_id, &[b.get_variable()]);
+            let mut shifted = res[1];
+            let new_bit = res[0];
+            if let Some(top_bit) = bit {
+                shifted = merge_byte_using_table::<_, _, 7>(cs, shifted, top_bit);
+            }
+            *b = UInt8 {
+                variable: shifted,
+                _marker: std::marker::PhantomData,
+            };
+            bit = Some(new_bit);
+        });
+        Self::from_le_bytes(cs, bytes)
+    }
+
+    /// Finds the result of multiplying `self` by `other` mod `modulo`.
```

Comment:
do mul through the a*b = q*P + rem

Reply:
Cannot quite understand what do you mean here, can you please clarify?

Reply:
If you are talking about leveraging the same approach as in U256, I would rather not do that because:

1. We don't actually use types u1024, u2048 and u4096. These were planned for modexp, but the decision was to go with u256. Since we've spent some time on these implementations, they were left.
2. There is enhanced complexity compared to U256, because `primitive_types` crate, from which we take the functionality for witnesses is limited by max U512 type, meaning that in these bigger types, there would be need to build complex compositional logics.

Summarizing, there is no particular benefit to spend time on building extra optimizations for unused things, which as well may introduce additional space for error.